test/framework/docker.go

```diff
@@ -61,3 +92,247 @@ func GetPublishedDockerPorts(
+	}
+
+	cmd := []string{"docker"}
+	args, err := formatDockerStopArgs(containers, options)
+	if err != nil {
+		return "", err
+	}
+	cmd = append(cmd, args...)
+	strings.Join(cmd, " ")
+
+	for _, container := range containers {
+		pfList, exists := u.portForwards[container]
+		if exists {
+			for _, vmPortForward := range pfList {
+				close(vmPortForward.StopChannel)
+			}
+			delete(u.portForwards, container)
+		}
+	}
+
+	return u.runDockerCommandInSession(t, "stop", strings.Join(cmd, " "), options.Logger)
+}
+
+func (u *RemoteDockerBackend) runDockerCommandInSession(t testing.TestingT, cmdName, cmd string, log *logger.Logger) (string, error) {
+	log.Logf(t, "Running 'docker %s' on remote host '%s', returning stdout: %q", cmdName, u.networking.RemoteHost.Address, cmd)
+	sshSession, err := u.networking.NewSession(path.Join("docker-commands", u.networking.RemoteHost.Address), cmdName, false, cmd)
+	if err != nil {
+		return "", err
+	}
+	err = sshSession.Run()
+	_ = sshSession.Close()
+	if err != nil {
+		return "", err
+	}
+
+	stdOut, err := os.ReadFile(sshSession.StdOutFile())
+	if err != nil {
+		return "", fmt.Errorf("failed to read stdout of docker %s command: %w", cmdName, err)
+	}
+	return string(stdOut), nil
+}
+
+func (u *RemoteDockerBackend) GetPublishedDockerPorts(
+	t testing.TestingT,
+	log *logger.Logger,
+	container string,
+	ports []uint32,
+) (map[uint32]uint32, error) {
+	listPortsCmd := []string{"docker", "port", container}
+	out, err := u.runDockerCommandInSession(t, "port", strings.Join(listPortsCmd, " "), log)
+	if err != nil {
+		// Sometimes the port may not be available immediately, and it can take some time.
+		// Since we didn't retry, tests were failing with and an error
+		// `missing port in address` on OSX.
+		time.Sleep(time.Millisecond * 500)
+		return nil, err
+	}
+
+	allPortMappings := strings.Split(out, "\n")
+
+	result := map[uint32]uint32{}
+	for _, port := range ports {
+		// example output format:
+		// 30094/tcp -> 0.0.0.0:40094
+		// 30094/tcp -> [::]:40094
+
+		mappingPrefix := fmt.Sprintf("%d/tcp")
+		var addresses []string
+		for _, portMapping := range allPortMappings {
+			if strings.HasPrefix(portMapping, mappingPrefix) {
+				parts := strings.Split(portMapping, " ")
+				addresses = append(addresses, parts[len(parts)-1])
+			}
+		}
+
+		if len(addresses) < 1 {
+			return nil, errors.Errorf("there are no addresses for port %d", port)
+		}
+		addr := addresses[0]
+		// on CircleCI (or other possible hosts), we get the ipv6 address in the format of ":::port",
+		// which is not parsable by the "net.SplitHostPort"
+		if strings.HasPrefix(addr, ":::") {
+			addr = "[::]:" + addr[3:]
+		}
+		_, pubPortStr, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, err
+		}
+
+		// setup SSH port forwarding for the remote port
+		pf, err := portForward(u.networking, fmt.Sprintf("%s:%d", u.networking.RemoteHost.Address, pubPortStr))
```

Comment:
net.JoinHostPort to work with ipv6

Reply:
Question is this actually the host public ip? We need to check this

Reply:
I'm going to use `127.0.0.1` here because the local port should be forwarded just to the remote host itself.